met-api/src/met_api/services/widget_timeline_service.py

```diff
@@ -28,23 +33,48 @@ def create_timeline(widget_id: int, timeline_details: dict):
         return widget_timeline
 
     @staticmethod
-    def update_timeline(widget_id: int, timeline_id: int, timeline_data: dict):
+    def update_timeline(widget_id: int, timeline_id: int, timeline_data: dict) -> Optional[WidgetTimelineModel]:
         """Update timeline widget."""
-        events = timeline_data.get("events")
+        events = timeline_data.get('events')
         first_event = events[0]
+
+        WidgetTimelineService._check_update_timeline_auth(first_event)
+
         widget_timeline: WidgetTimelineModel = WidgetTimelineModel.find_by_id(timeline_id)
-        authorization.check_auth(one_of_roles=(MembershipType.TEAM_MEMBER.name,
-                                               Role.EDIT_ENGAGEMENT.value), engagement_id=first_event.get('engagement_id'))
+
         if not widget_timeline:
-            raise KeyError('Timeline widget not found')
+            raise BusinessException(
+                error='Timeline widget not found',
+                status_code=HTTPStatus.BAD_REQUEST)
 
         if widget_timeline.widget_id != widget_id:
-            raise ValueError('Invalid widget ID')
+            raise BusinessException(
+                error='Invalid widget ID',
+                status_code=HTTPStatus.BAD_REQUEST)
 
         if widget_timeline.id != timeline_id:
-            raise ValueError('Invalid timeline ID')
+            raise BusinessException(
+                error='Invalid timeline ID',
+                status_code=HTTPStatus.BAD_REQUEST)
+
+        WidgetTimelineService._update_widget_timeline(widget_timeline, timeline_data)
+
+        return widget_timeline
 
-        return WidgetTimelineModel.update_timeline(timeline_id, timeline_data)
+    @staticmethod
+    def _check_update_timeline_auth(first_event):
+        eng_id = first_event.get('engagement_id')
+        authorization.check_auth(one_of_roles=(MembershipType.TEAM_MEMBER.name,
+                                               Role.EDIT_ENGAGEMENT.value), engagement_id=eng_id)
+
+    @staticmethod
+    def _update_widget_timeline(widget_timeline: WidgetTimelineModel, timeline_data: dict):
```

Comment:
Why was the update method for timelines moved to the service from the model? 

I've noticed a pattern in this app where models sometimes contain all CRUD methods and sometimes contain only some of them, with the service taking over. Do you happen to know the reason for this? In my mind, we'd strive to have all method handlers in the same location (perhaps the model). If there's a good reason for this that I'm missing, I'd be curious to hear it!

Reply:
We have intentionally designed our models to be minimalist, focusing on retrieving requested data. Not all models support full CRUD operations; for instance, tables like submissions are restricted from updates, and deletion is limited, typically reserved for entities like widgets. All business logic is centralized in the service layer. 
Moreover the logic in timeline was causing a cyclic dependency because we had the timeline event service being imported back to the model so I had to move that logic to the service.